pkg/controller/batchrelease/labelpatch/patcher.go

```diff
@@ -141,3 +149,45 @@ func calculateBatchReplicas(batches []v1beta1.ReleaseBatch, workloadReplicas, cu
 	}
 	return batchSize
 }
+
+type nativeDeploymentPatcher struct {
+	rp *realPatcher
+}
+
+// PatchPodBatchLabel patches label controller-revision-hash to pods owned by native deployments for pod-template-hash calculated
+// by rollouts differs from the one calculated by k8s. The hash calculated by rollouts is set as the value of controller-revision-hash
+func (p *nativeDeploymentPatcher) PatchPodBatchLabel(ctx *batchcontext.BatchContext) error {
+	if ctx.RolloutID == "" || len(ctx.Pods) == 0 {
+		return nil
+	}
+	pods := ctx.Pods
+	if ctx.FilterFunc != nil {
+		pods = ctx.FilterFunc(pods, ctx)
+	}
+	for i, pod := range pods {
+		if pod.Labels[v1.ControllerRevisionHashLabelKey] != "" {
+			continue
+		}
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil || owner.Kind != "ReplicaSet" {
+			klog.InfoS("Deployment Pod owner is not ReplicaSet", "pod", klog.KObj(pod), "rollout", p.rp.logKey)
+			continue
+		}
+		rs := &v1.ReplicaSet{}
+		if err := p.rp.Get(context.Background(), types.NamespacedName{Namespace: pod.Namespace, Name: owner.Name}, rs); err != nil {
+			klog.ErrorS(err, "Failed to get ReplicaSet", "pod", klog.KObj(pod), "rollout", p.rp.logKey)
+			return err
+		}
+		delete(rs.Spec.Template.ObjectMeta.Labels, v1.DefaultDeploymentUniqueLabelKey)
+		podClone := pod.DeepCopy()
+		hash := util.ComputeHash(&rs.Spec.Template, nil)
+		podClone.Labels[v1.ControllerRevisionHashLabelKey] = hash
+		by := fmt.Sprintf(`{"metadata":{"labels":{"%s":"%s"}}}`, v1.ControllerRevisionHashLabelKey, hash)
+		if err := p.rp.Patch(context.TODO(), podClone, client.RawPatch(types.StrategicMergePatchType, []byte(by))); err != nil {
```

Comment:
it is not necessary to deepclone the whole pod object, just call GetEmptyObjectWithKey